File: Cpp/Komputer.cpp
#pragma once
#include <iostream>
#include <vector>
#include "Cpu.cpp"
#include "Ram.cpp"
#include "Harddrive.cpp"
#include "Gpu.cpp"
#include "Motherboard.cpp"

using namespace std;

class Komputer {
private:
    string nama;
    Cpu cpu;
    Motherboard motherboard;
    Gpu gpu;
    vector<Ram> ramList;
    vector<Harddrive> storageList;

public:
    Komputer() {}

    Komputer(string nama, Cpu cpu, Motherboard mb, Gpu gpu) {
        this->nama = nama;
        this->cpu = cpu;
        this->motherboard = mb;
        this->gpu = gpu;
    }

    void setNama(string nama) {
        this->nama = nama;
    }

    void setCpu(Cpu cpu) {
        this->cpu = cpu;
    }

    void setMotherboard(Motherboard mb) {
        this->motherboard = mb;
    }

    void setGpu(Gpu gpu) {
        this->gpu = gpu;
    }

    void addRam(Ram ram) {
        ramList.push_back(ram);
    }

    void addStorage(Harddrive hdd) {
        storageList.push_back(hdd);
    }

    string getNama() {
        return this->nama;
    }

    Cpu getCpu() {
        return this->cpu;
    }

    Motherboard getMotherboard() {
        return this->motherboard;
    }

    Gpu getGpu() {
        return this->gpu;
    }

    vector<Ram> getRamList() {
        return this->ramList;
    }

    vector<Harddrive> getStorageList() {
        return this->storageList;
    }

    void displaySpecs() {
        cout << "=== " << nama << " ===\n";
        cout << "CPU: " << cpu.getNama() << " (" << cpu.getMerk() << ", " 
                << cpu.getJumlahCore() << " core, " << cpu.getKecepatanGHz() << " GHz)\n";
        cout << "Motherboard: " << motherboard.getNama() << " (" 
                << motherboard.getMerk() << ", Chipset: " << motherboard.getChipset() << ")\n";
        cout << "GPU: " << gpu.getNama() << " (" << gpu.getMerk() << ", " 
                << gpu.getVramGB() << " GB VRAM)\n";
    
        for (const Ram& ram : ramList) { // Gunakan const
            cout << "RAM: " << ram.getNama() << " (" << ram.getMerk() << ", " 
                    << ram.getKapasitasGB() << " GB)\n";
        }
    
        for (const Harddrive& hdd : storageList) { // Gunakan const
            cout << "Storage: " << hdd.getNama() << " (" << hdd.getMerk() << ", " 
                    << hdd.getKapasitasGB() << " GB, " << hdd.getTipe() << ")\n";
        }
    }
    

    ~Komputer() {}
};

File: Python/Komputer.py
class Komputer:
    def __init__(self, cpu, motherboard, gpu=None):
        self.__cpu = cpu
        self.__motherboard = motherboard
        self.__gpu = gpu  # Bisa None (Opsional)
        self.__ram_list = []  # List Komposisi RAM
        self.__harddrive_list = []  # List Komposisi Harddrive

    # Method untuk menambahkan RAM dan Harddrive
    def tambah_ram(self, ram):
        self.__ram_list.append(ram)

    def tambah_harddrive(self, harddrive):
        self.__harddrive_list.append(harddrive)

    def tampilkan_spesifikasi(self):
        print("===== Spesifikasi Komputer =====")
        print(f"CPU: {self.__cpu.get_nama()} ({self.__cpu.get_merk()}, {self.__cpu.get_jumlah_core()} Core, {self.__cpu.get_kecepatan_ghz()} GHz)")
        print(f"Motherboard: {self.__motherboard.get_nama()} ({self.__motherboard.get_merk()}, Chipset {self.__motherboard.get_chipset()})")

        if self.__gpu:
            print(f"GPU: {self.__gpu.get_nama()} ({self.__gpu.get_merk()}, {self.__gpu.get_memori_gb()} GB)")
        else:
            print("GPU: Tidak Ada")

        print("\nRAM:")
        for ram in self.__ram_list:
            print(f"- {ram.get_nama()} ({ram.get_merk()}, {ram.get_kapasitas_gb()} GB)")

        print("\nStorage:")
        for hdd in self.__harddrive_list:
            print(f"- {hdd.get_nama()} ({hdd.get_merk()}, {hdd.get_kapasitas_gb()} GB, {hdd.get_tipe()})")
